Enable tab complete for '-Format' parameter of Get-Date

iSazonov opened the pull request:
Add ArgumentCompletions attribute to '-Format' parameter of Get-Date to enable
IntelliSense (tab complete).

## Fix
1. Added ArgumentCompletions attribute to enable IntelliSense but allow not only named format values ("FileDate", "FileDateUniversal", "FileDateTime", "FileDateTimeUniversal") but also custom format strings ("YYYY-MM-dd").
2. Added a commit to replace tabs in test file.
3. Added new ArgumentCompletionsAttribute type which allow tab completions (IntelliSense) for parameters without throw on "invalid" values in contrast to ValidateSet.
4. Added tests for `-Format` parameter with named format values.
5. Added tests for ArgumentCompletionsAttribute

## Additional considerations
No tab completion test added for Get-Date because it seems we have no requirements to check tab completion for every parameter in all cmdlets.

iSazonov committed:
Enable tab complete for '-Format' parameter of Get-Date

Add ValidateSet attribute to '-Format' parameter of Get-Date to enable
IntelliSense (tab complete).

iSazonov committed:
Add 'AllowAll' named parameter in ValidateSet attribute

If AllowAll==true 'ValidateElement' always pass all elements.
Useful if we need enable IntelliSense (TabComplete) but effectively
disable validation
when a parameter can accept not only the listed values but also
arbitrary.
Ex.: Get-Date -Format

iSazonov committed:
Replace tabs with spaces

iSazonov committed:
Add tests for Get-Date -Format named values

vors commented:
Should we use another attribute name just for the IntelliSense scenario?
`[ValidateSet(AllowAll)]` looks cumbersome. We can inherit from `ValidateSetAttribute` and use some name that doesn't mention Validation. It's also not very clean, but at least for the end user it will be not as confusing imo.

iSazonov commented:
ValidateSet is a sealed.

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1564,6 +1564,14 @@ public sealed class ValidateSetAttribute : ValidateEnumeratedArgumentsAttribute
+        /// when a parameter can accept not only the listed values but also arbitrary.
+        /// Ex.: Get-Date -Format
+        /// </summary>
+        public bool AllowAll { get; set; } = false;
```
This doesn't feel right to me - you shouldn't use a `Validation` attribute if you aren't validating.

lzybkr replied:
This should be reverted.

lzybkr commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs:
```diff
@@ -206,6 +206,7 @@ public int Millisecond
         /// Unix format string
         /// </summary>
         [Parameter(ParameterSetName = "net")]
+        [ValidateSetAttribute("FileDate", "FileDateUniversal", "FileDateTime", "FileDateTimeUniversal", AllowAll=true)]
```
I would suggest using an `ArgumentCompleter` attribute instead - it's a bit more cumbersome, but if this scenario makes sense (and I think it probably does), we can find some way to make it simpler.

iSazonov replied:
Do you suggest enhance `ArgumentCompleter` to support `[ArgumentCompleter("FileDate", "FileDateUniversal", "FileDateTime", "FileDateTimeUniversal")] `?

lzybkr replied:
Just brainstorming here - but maybe a new attribute `ArgumentCompletions` that also implements `IArgumentCompleter` - this probably requires some additional code elsewhere, probably similar to the `ValidateSet` code for completion.

iSazonov replied:
A new attribute requires a lot of code. It seems easiest to add a new constructor `public ArgumentCompleterAttribute(params string[] completeStrings)`

lzybkr reviewed (changes requested):
Thanks for thinking about the usability here - I agree we should do something here, just not like this.

iSazonov committed:
Enhance ArgumentCompleter to support custom string array

Introduce new ArgumentCompleter constructor.
Invoke the constructor in NativeCommandArgumentCompletion()

iSazonov committed:
Use new ArgumentCompleterAttribute in Get-Date cmdlet

lzybkr commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs:
```diff
@@ -206,7 +206,7 @@ public int Millisecond
         /// </summary>
         [Parameter(ParameterSetName = "net")]
-        [ValidateSetAttribute("FileDate", "FileDateUniversal", "FileDateTime", "FileDateTimeUniversal", AllowAll=true)]
+        [ArgumentCompleter("FileDate", "FileDateUniversal", "FileDateTime", "FileDateTimeUniversal")]
```
This is much better, but I think still a little weird, I still think it's worth considering adding a new attribute `ArgumentCompletions`. It seems like you could take your class `StringArrayArgumentCompleter`, rename it, make it an attribute, and either duplicate a little bit of logic around `ValidateSet` or `ArgumentCompleter` in the tab completion code.

iSazonov replied:
Is your only thought that ArgumentCompleter is a strange name and you want to rename it?

lzybkr replied:
I wouldn't rename `ArgumentCompleter` even if it is a strange name (which I suppose it is.)

It's a really minor thing, but to me, `ArgumentCompleter` implies some dynamic code to compute the completions, whereas `ArgumentCompletions` is static.

powercode replied:
I have missed that for a long time - the ability to easily complete a set without limiting the parameter to the completions. 

Is it a reasonable compromise that you have to go IArgumentCompleter if you want to provide rich `CompletionResult`s, or should be enable this with `ArgumentCompletions` too?

iSazonov replied:
Done next iteration.

iSazonov committed:
Add ArgumentCompletionsAttribute

This attribute is used to specify an argument completions for a
parameter of a cmdlet or function without force throw unlike
ValidateSetAttribute.

lzybkr commented on src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs:
```diff
@@ -158,4 +158,56 @@ protected override void EndProcessing()
+            {
+                if (wordToCompletePattern.IsMatch(str))
+                {
+                    yield return new CompletionResult(str, str, CompletionResultType.Text, str);
```
Should this be `CompletionResultType.ParameterValue`? I think that's the only place the attribute will have any effect.

iSazonov replied:
I agree. 
Fixed.

lzybkr commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -2002,6 +2002,26 @@ internal static string ConcatenateStringPathArguments(CommandElementAst stringAs
+            var argumentCompletionsAttribute = parameter.CompiledAttributes.OfType<ArgumentCompletionsAttribute>().FirstOrDefault();
+            if (argumentCompletionsAttribute != null)
+            {
+                try
```
You are calling internal code which doesn't throw, so the try/catch is unnecessary.

iSazonov replied:
Fixed.

lzybkr commented on src/System.Management.Automation/engine/Attributes.cs:
```diff
@@ -1625,6 +1633,11 @@ protected override void ValidateElement(object element)
                             Metadata.ValidateNotNullFailure);
             }
 
+            if (AllowAll == true)
```
This should be reverted.

iSazonov replied:
Fixed.

lzybkr commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -1336,6 +1336,10 @@ private static Attribute NewValidateSetAttribute(AttributeAst ast)
                 {
                     result.IgnoreCase = s_attrArgToBoolConverter.Target(s_attrArgToBoolConverter, argValue);
                 }
+                else if (argumentName.Equals("AllowAll", StringComparison.OrdinalIgnoreCase))
```
This should be reverted.

iSazonov replied:
Fixed.

lzybkr reviewed (changes requested):
Just a little cleanup and I think it looks good.

iSazonov committed:
Remove try and change type

iSazonov committed:
Revert "Add 'AllowAll' named parameter in ValidateSet attribute"

This reverts commit 69363fd035807765b062f4ecff072d45bae6d3b1.

iSazonov committed:
Fix typo in comment.

lzybkr reviewed (approved):
Code changes are approved - but we should get a couple of tests for the new `ArgumentCompletions` attribute.

daxian-dbw commented:
@iSazonov Can you please update the PR description? Looks like the design has changed.

iSazonov committed:
Add tests

iSazonov commented:
Could you please help:  C# test failed on "ArgumentCompleter" abbreviation - full name ArgumentCompleterAttribute works. I cannot find how make "ArgumentCompleter" alias.

iSazonov committed:
Correct tests

daxian-dbw commented on test/powershell/Language/Parser/ExtensibleCompletion.Tests.ps1:
```diff
@@ -355,3 +355,97 @@ Describe "Additional type name completion tests" -Tags "CI" {
+    }
+
+    AfterAll {
+        #Remove-Module -ModuleInfo $testModule
```
Did you forget to uncomment this?

iSazonov replied:
Yes 😄 Today test is passed. I wonder - I don't change a code.

iSazonov committed:
Uncomment Remove-Module

iSazonov commented:
@lzybkr Thanks for great idea!
